### PR TITLE
Improve pivot tideman integration

### PR DESCRIPTION
**Pull request by carlschroedl**
Addresses #41 

This is a pretty rough cut that betrays my Laravel ignorance :) Could you guys please provide suggestions on how to better organize this code? I have only been doing manual testing via Postman so far. What would be a good approach for doing automated testing on this?

### CODE REVIEW

File: app/Http/Controllers/ResultController.php [hunk 1]
```diff
@@ -116,40 +116,20 @@ public function index(Election $election)
...
-        // TODO: populate ballots with actual user rankings
-        $ballot = new NBallot(1, ...$candidateLists);
-        $ballots = [$ballot];
-
-        // compute election results
-        $instance = new RankedPairsCalculator($tieBreaker);
-        $winnerOrder = $instance->calculate(sizeof($tidemanCandidates), ...$ballots);
+        $winnerOrder = $calculator->calculate($numWinners, ...$nBallots)->toArray();
 
+        $pivotCandidates = $election->candidates()->get()->keyBy('id');
```
Comment (tylerharter) [inline app/Http/Controllers/ResultController.php]: nice cleanup

File: app/Http/Controllers/ResultController.php [hunk 2]
```diff
@@ -116,40 +116,20 @@ public function index(Election $election)
...
-        foreach ($winnerOrder as $w) {
-            $candidate = $pivotCandidates[$w->getId()];
-            array_push($rv["order"], $candidate);
+        $orderedPivotCandidates = [];
+        //use explicit numerical indexing because order is very important
+        for($i = 0; $i < sizeof($winnerOrder); $i++) {
+            $tidemanCandidate = $winnerOrder[$i];
+            $candidateId = (int)$tidemanCandidate->getId();
+            $pivotCandidate = $pivotCandidates->get($candidateId);
+            $orderedPivotCandidates[] = $pivotCandidate;
```
Comment (tylerharter) [inline app/Http/Controllers/ResultController.php]: What is this syntax doing?  (for my information)

File: app/Http/Controllers/ResultController.php [hunk 3]
```diff
@@ -18,6 +20,80 @@ public function __construct()
...
+            });
+        })->toArray();
+        return $candidateRanksGroupedByElectorAndRank;
+    }
+
+    /**
+     * @param array of arrays of arrays as output by $this->groupRankingsByElectorAndRank
+     * @return array of NBallots
+     */
+    public function buildNBallots($candidateRanksGroupedByElectorAndRank)
```
Comment (tylerharter) [inline app/Http/Controllers/ResultController.php]: Do we need the NBallot concept?  There are N factorial different combos, so dedup at that level seems of limited value...

**Comment by carlschroedl** [comment]
Thanks for the speedy review!

File: app/Http/Controllers/ResultController.php (hunk 2 above)
Comment (carlschroedl) [reply app/Http/Controllers/ResultController.php]: Array append

File: app/Http/Controllers/ResultController.php (hunk 3 above)
Comment (carlschroedl) [reply app/Http/Controllers/ResultController.php]: Good point. I will add a task for simplifying.

Comment (carlschroedl) [reply app/Http/Controllers/ResultController.php]: https://github.com/pivot-libre/tideman/issues/56

File: app/Http/Controllers/ResultController.php (hunk 1 above)
Comment (carlschroedl) [reply app/Http/Controllers/ResultController.php]: 👍

File: app/Http/Controllers/ResultController.php [hunk 4]
```diff
@@ -18,6 +20,80 @@ public function __construct()
         $this->middleware('auth:api');
     }
 
+    /**
+     * @param int $electionId
+     * @return Illuminate\Collection of all CandidateRanks associated with
+     * the election identified by the parameterized  id.
+     */
+    public function getCandidateRankCollection($electionId)
```
Comment (carlschroedl) [inline app/Http/Controllers/ResultController.php]: Does this method belong in a Model? If so, Election or CandidateRank?

**Comment by carlschroedl** [comment]
Hi @aditya-cherukuri! Do you think this code conforms to Laravel's conventions on separation of concerns? Do the new methods all belong in the controller? Should any of them be moved to a util class or to a model?

**Pull request merged**